models/attachment.go

```diff
@@ -1,11 +1,149 @@
 package models
 
+import (
+	"archive/zip"
+	"bytes"
+	"encoding/base64"
+	"io"
+	"io/ioutil"
+	"path/filepath"
+	"strings"
+)
+
 // Attachment contains the fields and methods for
 // an email attachment
 type Attachment struct {
-	Id         int64  `json:"-"`
-	TemplateId int64  `json:"-"`
-	Content    string `json:"content"`
-	Type       string `json:"type"`
-	Name       string `json:"name"`
+	Id          int64  `json:"-"`
+	TemplateId  int64  `json:"-"`
+	Content     string `json:"content"`
+	Type        string `json:"type"`
+	Name        string `json:"name"`
+	vanillaFile bool   // Vanilla file has no template variables
+}
+
+// ValidateAttachment ensures that the provided attachment uses the supported template variables correctly.
+func (a Attachment) ValidateAttachment() error {
+
+	ptx := PhishingTemplateContext{
+		BaseRecipient: BaseRecipient{
+			FirstName: "Foo",
+			LastName:  "Bar",
+			Email:     "foo@bar.com",
+		},
+		BaseURL:     "http://testurl.com",
+		URL:         "http://testurl.com/?rid=1234567",
+		TrackingURL: "http://testurl.local/track?rid=1234567",
+		Tracker:     "<img alt='' style='display: none' src='http://testurl.local/track?rid=1234567'/>",
+		From:        "Foobus Barbus",
+		RId:         "1234567",
+	}
+
+	_, err := a.ApplyTemplate(ptx)
+	return err
+}
+
+// ApplyTemplate parses different attachment files and applies the supplied phishing template.
+func (a *Attachment) ApplyTemplate(ptx PhishingTemplateContext) (io.Reader, error) {
+
+	var processedAttachment string
+	decodedAttachment, err := base64.StdEncoding.DecodeString(a.Content) // Decode the attachment
```

Comment:
I'd love for us to try and keep things as an actual reader the whole way through if possible. I can't guarantee it'd work, but it'd be worth the exploration. I'll drop some comments throughout on where things might need to be adjusted.

For example, here we could make this `base64.NewDecoder(base64.StdEncoding, strings.NewReader(a.Content)`